cpp/marc_tools/analyze_non_standardized_keywords.cc

```diff
@@ -0,0 +1,144 @@
+/** \brief Utility for generating certain statistics for non-standardized keywords.
+ *  \author Dr. Johannes Ruscheinski (johannes.ruscheinski@uni-tuebingen.de)
+ *
+ *  \copyright 2020 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <algorithm>
+#include <unordered_map>
+#include <unordered_set>
+#include <cstdlib>
+#include "FileUtil.h"
+#include "MARC.h"
+#include "TextUtil.h"
+#include "util.h"
+
+
+namespace {
+
+
+std::string NormalizeKeyword(std::string keyword) {
+    TextUtil::CollapseAndTrimWhitespace(&keyword);
+    TextUtil::UTF8ToLower(&keyword);
+    return keyword;
+}
+
+
+void CollectNormalizedKeywordsAndTranslations(MARC::Reader * const reader,
+                                              std::unordered_set<std::string> * const normalized_keywords)
+{
+    unsigned record_count(0);
+    while (const auto record = reader->read()) {
+        ++record_count;
+
+        MARC::Record::KeywordAndSynonyms keyword_synonyms;
+        if (record.getKeywordAndSynonyms(&keyword_synonyms)) {
+            normalized_keywords->emplace(NormalizeKeyword(keyword_synonyms.getKeyword()));
+            for (const auto &keyword_synonym : keyword_synonyms)
+                normalized_keywords->emplace(NormalizeKeyword(keyword_synonym));
+        }
+    }
+
+    LOG_INFO("Processd " + std::to_string(record_count) + " authority records and found "
+             + std::to_string(normalized_keywords->size()) + " normalized keywords and their translations.");
+}
+
+
+const std::vector<std::string> non_normlized_keyword_tags{ "600", "610", "630", "648", "650", "651", "653", "655" };
+
+
+void ProcessField(const MARC::Record::Field &field, const std::unordered_set<std::string> &normalized_keywords,
+                  std::unordered_map<std::string, unsigned> * const unmatched_keywords_to_counts_map,
+                  unsigned * const matched_count, unsigned * const not_matched_count)
+{
+    const auto subfields(field.getSubfields());
+    for (const auto subfield_and_code : subfields) {
+        if (subfield_and_code.code_ != 'a')
+            continue;
```

Comment:
Impletentation in MARC.{h,cc} seems to address only fields 150,151,450. (cf. add_synonyms.cc line 392)